src/dune_engine/action_exec.ml

```diff
@@ -246,6 +246,29 @@ let rec exec t ~ectx ~eenv =
         )
       | exception _ -> Unix.symlink src dst);
     Fiber.return Done
+  | Hardlink (src, dst) ->
+    (* CR-someday amokhov: Instead of always falling back to copying, we could
+       detect if hardlinking works on Windows and if yes, use it. We do this in
+       the Dune cache implementation, so we can share some code. *)
+    (match Sys.win32 with
+    | true -> Io.copy_file ~src ~dst:(Path.build dst) ()
+    | false -> (
+      let rec follow_symlinks name =
+        let stats = Unix.lstat name in
+        match stats.st_kind with
+        | S_LNK ->
+          let link_name = Unix.readlink name in
+          let name = Filename.concat (Filename.dirname name) link_name in
+          follow_symlinks name
+        | _ -> name
+      in
```

Comment:
We could save one system call per symlink by calling `Unix.readlink` and catching `EINVAL`.

Reply:
Good idea, done!